common/achievement/create.ts

```diff
@@ -0,0 +1,72 @@
+import { Achievement_template } from '../../graphql/generated';
+import { prisma } from '../prisma';
+import { TemplateSelectEnum, getAchievementTemplates } from './template';
+import { UserAchievementContext } from './types';
+
+async function doesUserAchievementAlreadyExist(templateId: number, userId: string, context?: UserAchievementContext) {
+    // TODO - check if user achievement exist for one match or one subcourse
+    const userAchievement = await prisma.user_achievement.findFirst({
+        where: { templateId, userId },
+        select: { id: true, userId: true, achievedAt: true, context: true, template: true },
+    });
+    if (!userAchievement) {
+        return false;
+    }
+    return userAchievement;
+}
+
+async function getOrCreateUserAchievement(template: Achievement_template, userId: string, context?: UserAchievementContext) {
+    const existingUserAchievement = await doesUserAchievementAlreadyExist(template.id, userId, context);
+    if (existingUserAchievement === false) {
+        return await createUserAchievement(template, userId, context);
+    }
+    return existingUserAchievement;
+}
+
+async function createUserAchievement(templateToCreate: Achievement_template, userId: string, context: UserAchievementContext) {
+    switch (templateToCreate.type) {
+        case 'SEQUENTIAL':
+            return await createSequentialAchievement(templateToCreate, userId, context);
+            break;
+        case 'TIERED':
+            // await createTieredAchievement(userAchievements, userId, context);
+            break;
+        case 'STREAK':
+            // await createStreakAchievement(userAchievements[0], userId, context);
+            break;
+        default:
+            console.log('DEFAULT');
+    }
+}
+
+async function createSequentialAchievement(templateToCreate: Achievement_template, userId: string, context: UserAchievementContext) {
+    const templatesByGroup = await getAchievementTemplates(TemplateSelectEnum.BY_GROUP);
+    const userAchievementsByGroup = await prisma.user_achievement.findMany({
+        where: { template: { group: templateToCreate.group } },
+        // orderBy: { template: { groupOrder: 'asc' } },
+    });
+
+    // const lastStepIndex = userAchievementsByGroup.length > 0 ? templateToCreate.groupOrder : 0;
+    // const nextStepIndex = lastStepIndex + 1;
+    const nextStepIndex = userAchievementsByGroup.length > 0 ? templateToCreate.groupOrder + 1 : 1;
+
+    const templatesForGroup = templatesByGroup.get(templateToCreate.group);
+    if (templatesForGroup) {
+        const nextStepTemplate = templatesForGroup.find((template) => template.groupOrder === nextStepIndex);
+        const isLastStepInGroup = nextStepIndex === templatesForGroup.length;
```

Comment:
soo for some sequential achievements - i.e. "Neue Lernunterstützung" - the first step is a match request. So at this point we don't know the `matchId`. On step two we need the `matchId`, because we need the `matchpartner name`(see photo) in the subtitle. So its easy to pass the correct `context` on the action that triggers the achievement. So if we create the whole group at once, we have to make a update on the achievement, but in our concept we work with the creation of the next step/stage by copying the next template. 

![Bildschirmfoto 2023-11-28 um 16 27 23](https://github.com/corona-school/backend/assets/115979086/6cc31b28-8be4-4230-ac1a-d4370f23a56d)

Reply:
Thank you for the example. Everything makes sense now 🙂 

Could you please add this explanation above her?
https://github.com/corona-school/backend/pull/898/files#diff-8576be384b0630de8bf69b5c167f22931dff3c1cfb254a9d3b07df009cd74244R103

Reply:
A comment was added to explain the process with an example:
```
    // Here a user template is created for the next template in the group. This is done to always have the data availible for the next step.
    // This could mean to, for example, have the name of a match partner that is not yet availible due to a unfinished matching process.
```